### PR TITLE
[JENKINS-45245] Work around failure of IDEA to add jenkins-war.war to the test CP

### PR DESCRIPTION
**Pull request by jglick**
Possible fix of [JENKINS-45245](https://issues.jenkins-ci.org/browse/JENKINS-45245). Not tested with actual IDEA, since I do not use IDEA and would not even know how to set it up to see the problem to begin with. But does the right thing if `"/winstone.jar"` is temporarily mangled to something nonexistent.

@reviewbybees @jenkinsci/code-reviewers

### CODE REVIEW

**Comment by jglick** [comment]
BTW to test this locally you should be able to just

```xml
<jenkins-test-harness.version>2.27-20170919.194540-1</jenkins-test-harness.version>
```

assuming IDEA at least supports timestamped snapshots (perhaps after an initial `mvn validate` priming).

File: src/main/java/org/jvnet/hudson/test/WarExploder.java [hunk 1]
```diff
@@ -74,18 +78,31 @@ private static File explode() throws Exception {
...
         URL winstone = WarExploder.class.getResource("/winstone.jar");
-        if(winstone==null)
-            // impossible, since the test harness pulls in jenkins.war
-            throw new AssertionError("jenkins.war is not in the classpath. If you are running this from the core workspace, run 'mvn install' to create the war image in war/target/jenkins. If running from a plugin, try using `mvn clean test` (cf. JENKINS-45245).");
-        File war = Which.jarFile(Class.forName("executable.Executable"));
+        if (winstone != null) {
+            war = Which.jarFile(Class.forName("executable.Executable"));
+        } else {
+            // JENKINS-45245: work around incorrect test classpath in IDEA. Note that this will not correctly handle timestamped snapshots; in that case use `mvn test`.
+            File core = Which.jarFile(Jenkins.class); // will fail with IAE if have neither jenkins-war.war nor jenkins-core.jar in ${java.class.path}
```
Comment (abayer) [inline src/main/java/org/jvnet/hudson/test/WarExploder.java]: s/IAE/IDEA/

**Comment by abayer** [comment]
Double-checked with a different plugin than I'd been doing experiments with, and yup, looks good.

File: src/main/java/org/jvnet/hudson/test/WarExploder.java (hunk 1 above)
Comment (jglick) [reply src/main/java/org/jvnet/hudson/test/WarExploder.java]: Nope, `IllegalArgumentException`.

Comment (abayer) [reply src/main/java/org/jvnet/hudson/test/WarExploder.java]: Might be worth spelling that out, then, 'cos `IDEA` a line above `IAE` is, as demonstrated here, potentially confusing. =)

**Comment by jglick** [comment]
Let us give some other folks a chance to test/review, but if it still looks good, we should be able to cut a JTH release tomorrow-ish (I would not mind getting #70 & #71 in too) and integrate into `plugin-pom`.

File: src/main/java/org/jvnet/hudson/test/WarExploder.java (hunk 1 above)
Comment (svanoort) [reply src/main/java/org/jvnet/hudson/test/WarExploder.java]: Seconded.

File: src/main/java/org/jvnet/hudson/test/WarExploder.java [hunk 2]
```diff
@@ -74,18 +78,31 @@ private static File explode() throws Exception {
...
+        } else {
+            // JENKINS-45245: work around incorrect test classpath in IDEA. Note that this will not correctly handle timestamped snapshots; in that case use `mvn test`.
+            File core = Which.jarFile(Jenkins.class); // will fail with IllegalArgumentException if have neither jenkins-war.war nor jenkins-core.jar in ${java.class.path}
+            String version = core.getParentFile().getName();
+            if (core.getName().equals("jenkins-core-" + version + ".jar") && core.getParentFile().getParentFile().getName().equals("jenkins-core")) {
+                war = new File(new File(new File(core.getParentFile().getParentFile().getParentFile(), "jenkins-war"), version), "jenkins-war-" + version + ".war");
+                if (!war.isFile()) {
+                    throw new AssertionError(war + " does not yet exist. Prime your development environment by running `mvn validate`.");
+                }
+                LOGGER.log(Level.INFO, "{0} is the continuation of the classpath by other means", war);
```
Comment (svanoort) [inline src/main/java/org/jvnet/hudson/test/WarExploder.java]: 🐜  Not a terribly meaningful log message and probably not necessary to most users.  I think what you're trying to say is something like "Loaded Jenkins core classes from JAR {0}" -- probably at a CONFIG or FINE log level, just for specialty debugging.

File: src/main/java/org/jvnet/hudson/test/WarExploder.java [hunk 3]
```diff
@@ -74,18 +78,31 @@ private static File explode() throws Exception {
...
+        File war;
         URL winstone = WarExploder.class.getResource("/winstone.jar");
-        if(winstone==null)
-            // impossible, since the test harness pulls in jenkins.war
-            throw new AssertionError("jenkins.war is not in the classpath. If you are running this from the core workspace, run 'mvn install' to create the war image in war/target/jenkins. If running from a plugin, try using `mvn clean test` (cf. JENKINS-45245).");
-        File war = Which.jarFile(Class.forName("executable.Executable"));
+        if (winstone != null) {
+            war = Which.jarFile(Class.forName("executable.Executable"));
+        } else {
+            // JENKINS-45245: work around incorrect test classpath in IDEA. Note that this will not correctly handle timestamped snapshots; in that case use `mvn test`.
```
Comment (daniel-beck) [inline src/main/java/org/jvnet/hudson/test/WarExploder.java]: Unclear what this means. Will a `jenkins.version` snapshot dependency break with this? If so, is there a way to detect it here and output this comment as an explicit error message (perhaps as part of the assertion error on 99?)

**Comment by hoshsadiq** [comment]
Seems to work for me

File: src/main/java/org/jvnet/hudson/test/WarExploder.java (hunk 3 above)
Comment (jglick) [reply src/main/java/org/jvnet/hudson/test/WarExploder.java]: It just means that given a `jenkins-core.version` of `2.99-20180101.1234` or whatever and `jenkins-war.version` `2.99-20180101.5678`, from the perspective of `WarExploder` we only know that we are running out of `…/jenkins-core/2.99-SNAPSHOT/jenkins-core-2.99-SNAPSHOT.jar` and thus that we should load `…/jenkins-war/2.99-SNAPSHOT/jenkins-war-2.99-SNAPSHOT.war`. It does not necessarily know the specific version to use.

Basically, a locally built snapshot dep should be fine, but do not expect a checkout of a downstream plugin PR using an upstream core PR to allow you to run tests from IDEA without using `mvn test`. Which you would rarely want to do anyway, since if you were running tests locally, it is probably because you are testing local changes including to core.

Comment (jglick) [reply src/main/java/org/jvnet/hudson/test/WarExploder.java]: Of course all of this is speculative—I cannot myself try it.

Comment (daniel-beck) [reply src/main/java/org/jvnet/hudson/test/WarExploder.java]: Perhaps `if (version.contains("SNAPSHOT")) { /* log something saying we might pick up the wrong war *? }`? But this can go ahead without that, could always be added later.

Comment (jglick) [reply src/main/java/org/jvnet/hudson/test/WarExploder.java]: Yeah I would leave this to an IDEA user actually trying to use snapshot core deps to further test & refine.

File: src/main/java/org/jvnet/hudson/test/WarExploder.java [hunk 4]
```diff
@@ -107,9 +124,12 @@ private static File explode() throws Exception {
...
             new File(explodeDir + ".exploding").delete();
         } else {
-            System.out.println("Picking up existing exploded jenkins.war at "+explodeDir.getAbsolutePath());
+            LOGGER.log(Level.INFO, "Picking up existing exploded jenkins.war at {0}", explodeDir.getAbsolutePath());
         }
 
         return explodeDir;
     }
+
+    private WarExploder() {}
```
Comment (oleg-nenashev) [inline src/main/java/org/jvnet/hudson/test/WarExploder.java]: Unrelated, it may probably cause issues in some custom implementations which instantiate War Exploder due to whatever reason.

I suggest reverting it before merging

File: src/main/java/org/jvnet/hudson/test/WarExploder.java (hunk 3 above)
Comment (oleg-nenashev) [reply src/main/java/org/jvnet/hudson/test/WarExploder.java]: Fine for me.

File: src/main/java/org/jvnet/hudson/test/WarExploder.java (hunk 4 above)
Comment (jglick) [reply src/main/java/org/jvnet/hudson/test/WarExploder.java]: It consists only of a `static` method. There is no reason to ever instantiate this class. That is the purpose of the `private` constructor—to clarify this.

Anyway we do not guarantee binary compatibility for test harnesses. Upgrades are done only by developers.
